lib/cPanel/PublicAPI.pm

```diff
@@ -305,129 +298,68 @@ sub api_request {
-                  $self->{'usessl'}
-                  ? IO::Socket::SSL->new(@connection_args)
-                  : IO::Socket::INET->new($connection_string);
-                if ( !$whm_sock ) {
-                    undef $whm_sock;
-                    $self->error("Could not connect to $connection_string: $!");
-                    die $self->{'error'};    #exit eval
-                }
-                else {
-                    $whm_sock_host = $connection_string;
-                }
-            }
-            my $connection_type = ( $keepalive ? 'Keep-Alive' : 'Close' );
 
-            $headers = $self->format_http_headers($headers) if ref $headers;
-            if ( $method eq 'POST' || $method eq 'PUT' ) {
-                $formdata = $self->format_http_query($formdata) if ref $formdata;
-                print {$whm_sock} "$method $uri HTTP/1.1\r\nHost: $server_name\r\nConnection: $connection_type\r\n" . 'Content-Length: ' . length($formdata) . "\r\n" . "User-Agent: " . 'cPanel::PublicAPI (perl) ' . $VERSION . "\r\n" . "Authorization: $authstr\r\n$headers\r\n" . $formdata;
-                $self->debug( ( "$method $uri HTTP/1.1\r\nHost: $server_name\r\nConnection: $connection_type\r\n" . 'Content-Length: ' . length($formdata) . "\r\n" . "User-Agent: " . 'cPanel::PublicAPI (perl) ' . $VERSION . "\r\n" . "Authorization: $authstr\r\n$headers\r\n" . $formdata ) ) if $self->{'debug'};
-            }
-            else {
-                print {$whm_sock} "$method "
-                  . ( $uri . '?' . ( ref $formdata ? join( '&', map { $CFG{'uri_encoder_func'}->($_) . '=' . $CFG{'uri_encoder_func'}->( $formdata->{$_} ) } keys %{$formdata} ) : $formdata ) )
-                  . " HTTP/1.1\r\nHost: $server_name\r\nConnection: $connection_type\r\n"
-                  . "User-Agent: "
-                  . 'cPanel::PublicAPI (perl) '
-                  . $VERSION . "\r\n"
-                  . "Authorization: $authstr\r\n$headers\r\n";
-
-                $self->debug( "$method " . ( $uri . '?' . ( ref $formdata ? join( '&', map { $CFG{'uri_encoder_func'}->($_) . '=' . $CFG{'uri_encoder_func'}->( $formdata->{$_} ) } keys %{$formdata} ) : $formdata ) ) . " HTTP/1.1\r\nHost: $server_name\r\nConnection: $connection_type\r\n" . "User-Agent: " . 'cPanel::PublicAPI (perl) ' . $VERSION . "\r\n" . "Authorization: $authstr\r\n$headers\r\n" )
-                  if $self->{'debug'};
+        $formdata = $self->format_http_query($formdata) if ref $formdata;
 
+        my $scheme = $self->{'usessl'} ? "https" : "http";
+        my $url = "$scheme://$remote_server:$port$uri";
+        my $content;
+        if ( $method eq 'POST' || $method eq 'PUT' ) {
+            $content = $formdata;
+        }
+        else {
+            $url .= "?$formdata";
+        }
+        if ( !ref $headers ) {
+            my @lines = split /\r\n/, $headers;
+            $headers = {};
+            foreach my $line (@lines) {
+                next unless length $line;
```

Comment:
An empty line indicates the end of headers.  Should this be `last unless length $line;`?